test/apps/iodemo/io_demo.cc

```diff
@@ -248,16 +330,56 @@ class P2pDemoCommon : public UcxContext {
 
     class Buffer {
     public:
-        Buffer(size_t size, MemoryPool<Buffer, true>& pool) :
-            _capacity(size), _buffer(memalign(ALIGNMENT, size)), _size(0),
-            _pool(pool) {
+        Buffer(size_t size, MemoryPool<Buffer, true> &pool,
+               ucs_memory_type_t memory_type = UCS_MEMORY_TYPE_HOST) :
+            _capacity(size), _size(0), _pool(pool), _memory_type(memory_type)
+        {
+#ifdef HAVE_CUDA
+            cudaError_t cerr;
+#endif
+            switch (memory_type) {
+#ifdef HAVE_CUDA
+            case UCS_MEMORY_TYPE_CUDA:
+                cerr = cudaMalloc(&_buffer, size);
+                if (cerr != cudaSuccess) {
+                    _buffer = NULL;
+                }
+                break;
+            case UCS_MEMORY_TYPE_CUDA_MANAGED:
+                cerr = cudaMallocManaged(&_buffer, size, cudaMemAttachGlobal);
+                if (cerr != cudaSuccess) {
+                    _buffer = NULL;
+                }
+                break;
+#endif
+            case UCS_MEMORY_TYPE_HOST:
+                _buffer = memalign(ALIGNMENT, size);
+                break;
```

Comment:
maybe split up for different classes?
For example:
```
          Buffer
         /      \ 
HostBuffer      CudaBuffer
```
and implement mem type fabric in base class:
```
switch(memory_type) {
case UCS_MEMORY_TYPE_HOST:
    Buffer::_impl = new HostBuffer(...)
...
```